cibuildwheel/macos.py

```diff
@@ -180,6 +240,35 @@ def setup_python(python_configuration: PythonConfiguration,
         # https://github.com/python/cpython/blob/a5ed2fe0eedefa1649aa93ee74a0bafc8e628a10/Lib/_osx_support.py#L260
         env.setdefault('ARCHFLAGS', '-arch x86_64')
 
+    if python_configuration.version == '3.9':
+        if python_configuration.identifier.endswith('x86_64'):
+            # even on the macos11.0 Python installer, on the x86_64 side it's
+            # compatible back to 10.9.
+            env.setdefault('_PYTHON_HOST_PLATFORM', 'macosx-10.9-x86_64')
+            env.setdefault('ARCHFLAGS', '-arch x86_64')
+        elif python_configuration.identifier.endswith('arm64'):
+            # macOS 11 is the first OS with arm64 support, so the wheels
+            # have that as a minimum.
+            env.setdefault('_PYTHON_HOST_PLATFORM', 'macosx-11.0-arm64')
+            env.setdefault('ARCHFLAGS', '-arch arm64')
+        elif python_configuration.identifier.endswith('universal2'):
+            if get_macos_version() < (10, 16):
+                # we can do universal2 builds on macos 10.15, but we need to
+                # set ARCHFLAGS otherwise CPython sets it to `-arch x86_64`
+                env.setdefault('ARCHFLAGS', '-arch arm64 -arch x86_64')
+
+    if python_configuration.identifier.endswith('arm64') or python_configuration.identifier.endswith('universal2'):
+        if get_macos_version() < (10, 16):
+            # xcode 12.2 or higher can build arm64 on macos 10.15 or below, but
+            # needs the correct SDK selected
+
+            # however, different versions of Xcode contain different SDK
+            # versions...
+            if get_xcode_version() == (12, 2):
+                env.setdefault('SDKROOT', 'macosx11.0')
+            else:
+                env.setdefault('SDKROOT', 'macosx11.1')
```

Comment:
@joerick, 
Won't this be failing in the near future ? I guess the only way is to update this each time there's a new version of Xcode / macosx SDK ?

Reply:
Unfortunately, yes, it's probably gonna need a few updates, if Apple release updated Xcodes with newer SDKs. Normally there are about [4 SDK releases per Xcode major version](https://xcodereleases.com/)...

I suppose another way to do this would be to parse the output of `xcodebuild -showsdks`, which looks like this:

```shell
$ xcodebuild -showsdks 2>/dev/null
iOS SDKs:
	iOS 14.3                      	-sdk iphoneos14.3

iOS Simulator SDKs:
	Simulator - iOS 14.3          	-sdk iphonesimulator14.3

macOS SDKs:
	DriverKit 20.2                	-sdk driverkit.macosx20.2
	macOS 11.1                    	-sdk macosx11.1

tvOS SDKs:
	tvOS 14.3                     	-sdk appletvos14.3

tvOS Simulator SDKs:
	Simulator - tvOS 14.3         	-sdk appletvsimulator14.3

watchOS SDKs:
	watchOS 7.2                   	-sdk watchos7.2

watchOS Simulator SDKs:
	Simulator - watchOS 7.2       	-sdk watchsimulator7.2
```

Kind of messy output, but perhaps we could look for `-sdk macosx(\d+)\.[\.\d]+` and choose anything >11...

Reply:
Up to you.
On one hand, doing nothing now will require some updates now & then.
On the other hand, doing something requires adding some parsing code...

> Kind of messy output, but perhaps we could look for -sdk macosx(\d+)\.[\.\d]+ and choose anything >11...

That might not even add that much code. Maybe that's the good option. Up-to-you.

Reply:
I think pulling it this way would be more reliable and less work long term than manual updates.

Reply:
Yeah, maybe it would be better to detect the SDK. I'm not too worried about the amount of code, more that we're parsing something that doesn't look particularly designed for machine-readability and might change..? maybe there's a better way to list the sdks?

Reply:
This would be an alternative, but it's not documented...
```python
import subprocess
import plistlib
from pathlib import Path

xcode_path = Path(subprocess.check_output(['xcode-select', '--print-path'], text=True).strip())
sdk_settings_plist = xcode_path / 'Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/SDKSettings.plist'
sdk_settings = plistlib.loads(sdk_settings_plist.read_bytes())
print(sdk_settings['CanonicalName'])
# prints 'macosx11.1'
```

But, on reflection, looking around the web, `xcodebuild -showsdks` hasn't changed its format in at least 8 years, and it's the documented way to list sdks, so I think I'll do that.

Reply:
See 74170ba for implementation.